File: frontend/lib/data/models/nearest_stop_result.dart
class NearestStopResult {
  final String buildingName;
  final double buildingLatitude;
  final double buildingLongitude;
  final String busStopName;
  final String busStopDisplayName;
  final double busStopLatitude;
  final double busStopLongitude;
  final double distanceMeters;

  const NearestStopResult({
    required this.buildingName,
    required this.buildingLatitude,
    required this.buildingLongitude,
    required this.busStopName,
    required this.busStopDisplayName,
    required this.busStopLatitude,
    required this.busStopLongitude,
    required this.distanceMeters,
  });

  factory NearestStopResult.fromJson(Map<String, dynamic> json) =>
      NearestStopResult(
        buildingName: json['buildingName'] as String? ?? '',
        buildingLatitude: (json['buildingLatitude'] as num?)?.toDouble() ?? 0,
        buildingLongitude: (json['buildingLongitude'] as num?)?.toDouble() ?? 0,
        busStopName: json['busStopName'] as String? ?? '',
        busStopDisplayName: json['busStopDisplayName'] as String? ?? '',
        busStopLatitude: (json['busStopLatitude'] as num?)?.toDouble() ?? 0,
        busStopLongitude: (json['busStopLongitude'] as num?)?.toDouble() ?? 0,
        distanceMeters: (json['distanceMeters'] as num?)?.toDouble() ?? 0,
      );
}

File: backend/src/main/java/com/nusmotion/backend/dto/NearestStopResult.java
package com.nusmotion.backend.dto;

/**
 * Response DTO for the "nearest bus stop to a building" endpoint.
 *
 * LEARNING NOTE:
 * - This record is NOT mapped from an upstream API — we CREATE it ourselves.
 *   The service layer computes the nearest stop using the Haversine formula,
 *   then assembles this record from building + bus stop data.
 * - We include both the building and the bus stop coordinates so the Flutter
 *   frontend can draw a line between them on the map.
 * - distanceMeters tells the user how far the walk is (straight-line distance).
 */
public record NearestStopResult(
        String buildingName,
        double buildingLatitude,
        double buildingLongitude,
        String busStopName,
        String busStopDisplayName,
        double busStopLatitude,
        double busStopLongitude,
        double distanceMeters
) {}
